### ISSUE TITLE
[BUG] Feed is broken (Fixed in pull request #698)

### ISSUE THREAD

**Issue by TheHCJ** [jonjomckay/fritter#668]
# I have now fixed this, check out https://github.com/jonjomckay/fritter/pull/698
[Github Releases](https://github.com/TheHCJ/fritter/releases)
[Source Code](https://github.com/TheHCJ/fritter/)

## Comment on feeds
https://github.com/jonjomckay/fritter/issues/668#issuecomment-1528784190

**Describe the bug**
I get the error `type 'Null' is not a subtype of type 'String' in type cast` when refreshing my feed

**To Reproduce**
1. Go to 'Feed'
2. Click on 'Refresh'
3. See error

**Screenshots**
https://user-images.githubusercontent.com/123662124/233856792-46a33b7b-c09d-47f1-9c06-25a1d773e3ed.png

**Device**
Model: Moto G62 5G
OS: Android 12 (API 31)
Locale: en
Version: 3.0.0-beta7+300000957
Flavor: Github

**Comment by moocow1452** [comment]
Same, it was working earlier for me, now it's not.

**Comment by shayne3717** [comment]
I was able to get it working again by unchecking the "Show Replies" option in the overflow menu; it seems like unchecking either that or "Show Retweets" fixes the issue. Checking both causes the issue to return.

**Comment by moocow1452** [comment]
That's not a consistent fix on my side, error returns after a refresh, and no amount of toggling will bring the feed back.

**Comment by AndiDink00x** [comment]
Same thing here. Again app became unusable because of this bug.

**Comment by d3vv3** [comment]
I get “Twitter says access to this is forbidden”, and only a “Back” button appears.

**Comment by anton-isidore** [comment]
This happens practically immediately after closing the app. 

I assume something bad gets saved into local storage that keeps crashing the app.


Sadly this bug renders the app practically unusable.

**Comment by hounsell** [comment]
Having a quick look into this issue, it looks like the issue is that the top cursor is missing in some stored chunks, which trips this up:
https://github.com/jonjomckay/fritter/blob/69bdee9d35dbf332afaad455e4ace724e39f95c7/lib/group/_feed.dart#L148-L150

I've worked around this in my local copy just by adding in an additional null check for cursor_top before trying to cast it, like so:

```dart
if (latestChunk != null && latestChunk['cursor_top'] != null) {
    searchCursor = latestChunk['cursor_top'] as String;
} else {
```

And this does seem to get the app functioning again, but now some tweets in my feed get duplicated. I guess this is because the chunks are getting re-fetched, or at least another chunk is requested that contains some of the same tweets because the cursor wasn't in place to delineate where the existing chunk finished off. The proper fix is probably then to look at why some stored chunks are missing their top cursor.

**Comment by TheHCJ** [comment]
> Having a quick look into this issue, it looks like the issue is that the top cursor is missing in some stored chunks, which trips this up:
> 
> https://github.com/jonjomckay/fritter/blob/69bdee9d35dbf332afaad455e4ace724e39f95c7/lib/group/_feed.dart#L148-L150
> 
> I've worked around this in my local copy just by adding in an additional null check for cursor_top before trying to cast it, like so:
> 
> ```dart
> if (latestChunk != null && latestChunk['cursor_top'] != null) {
>     searchCursor = latestChunk['cursor_top'] as String;
> } else {
> ```
> 
> And this does seem to get the app functioning again, but now some tweets in my feed get duplicated. I guess this is because the chunks are getting re-fetched, or at least another chunk is requested that contains some of the same tweets because the cursor wasn't in place to delineate where the existing chunk finished off. The proper fix is probably then to look at why some stored chunks are missing their top cursor.
Thanks for this info! I am now testing your code out to find a fix 😀

**Comment by j-fbriere** [comment]
This #679 PR should fix the problem.

**Comment by moocow1452** [comment]
Not sure if that also fixes this new issue. [Screenshot_20230427-003347.png](https://user-images.githubusercontent.com/1383448/234760184-0698869f-d231-457d-b625-63d17644ca37.png)

**Comment by j-fbriere** [comment]
It doesn't fix the new issue that appeared today unfortunately.
It seems that the graphql endpoint has been shut down. 
At least for the moment.
Since a few hours ago, it returns a 404 error.

**Comment by jonjomckay** [comment]
I'm sorry to say it, but I think this is probably going to be the end of feed support in Fritter. It was always a fragile hack on top of the search API, and now it looks like Muskrat has finally removed support for searching publicly.

I don't see any way around this, besides logging in with an account, which is something Fritter won't support by design, due to the privacy concerns, and the point of Fritter is to browse Twitter as invisibly as possible.

Before people jump in and say "why can't you combine multiple profile tweet feeds into one, and use that?" - it's not feasible, unfortunately. Only 20 tweets can be fetched at a time, from a single profile, and there are now heavy rate limits in place for those endpoints. Even having a feed of 10 users who tweet semi-regularly would encounter constant rate limit errors, and be incredibly slow to load.

I'll shortly be removing the code that handles and displays feeds, and will be modifying groups to provide a list of accounts, rather than a feed view. I know that's going to be bad news for some of you, but I can't leave the functionality in and broken.

On the plus side, supporting the feed was always where the majority of my time was spent while maintaining Fritter, and it was always the most frustrating part to work on, due to its fragility. Not having to maintain it any more will allow me to better utilise the time I spend on Fritter to work on the other parts of the app 🙂 

Thanks for understanding, and I hope you all continue to use and enjoy the app, until Elon decides to remove even more public endpoints 🙈

**Comment by Mrnofish** [comment]
@jonjomckay before you get rid of the feed altogether, have you taken a look at nitter ?

There is some talk about using unofficial tokens, which might be worth looking into.

Another alternative is operating a bunch of accounts to circumvent the search ban.

This latter approach, as I understand it, would work more or less like Aurora Store in anonymous mode. Not perfect, I know.

Since the nitter folks seem intent on not getting shut out, that project might be also an avenue to keep in mind.

It should already be possible to recreate the feed by fetching RSS feeds for individual accounts and then aggregating them.

I understand of course that even looking into this is a big ask, with Mr "Free" Speech at the helm it was always going to be an uphill fight, but I cannot think of a better way of flipping the bird, other than finding ways to keep using the service on our terms rather than head honcho's.

**Comment by jonjomckay** [comment]
Yep, I've been following the threads on the Nitter project, but the solutions won't fit for Fritter.

* Fritter was already using these "unofficial tokens" previously, and they no longer work for the search API, nor for the GraphQL endpoints, which both Nitter and Fritter moved to
* I simply don't have the time or financial resources to operate a service that rotates a bunch of (continously-banned or rate limited) accounts, and this would go against the privacy-first stance that Fritter stands for, by having to proxy all requests through a third-party service (even if it is operated by the me, the maintainer)
* Nitter's RSS feeds don't contain nearly enough information to populate Fritter's UI 

I'm open to solutions, but I'm only able to commit a small amount of time to Fritter, so any solutions will need to be pretty much maintenance-free, which I don't hold much hope for.

**Comment by Mrnofish** [comment]
Perhaps I could've been clearer - what I meant to say is that a client app could also plug into a Nitter instance, leaving the task of managing the sacrificial accounts to the instances' admins.

Such app could rotate through several such instances and ensure no one operator enjoys a gatekeeper position on the user's data, or allow the user to pick an instance that they trust.

There are several apps that have done something along the lines for other privacy-oriented frontends, covering e.g. Reddit, YouTube and other major services.

Again, not asking that you personally do something, I can certainly appreciate how much thankless work goes into even the simplest of projects.

You've done an excellent job with Fritter, and I've enjoyed using it - thank you so much for everything that you've done so far.

**Comment by geominorai** [comment]
Thanks for keeping up with this issue @jonjomckay despite its unfortunate conclusions.

If the restrictive rate limits for building a real-time feed is impractical, perhaps a workaround and compromise would be to build a feed by caching tweets from individual profiles queried over time (with some configurable storage limit, e.g. 3 months) as if intending for offline viewing and this cache updates when manually visiting the profile or during a global update. This global update of all subscribed profiles on the feed could be distributed over the course of the day as a background task for the app.

This is similar in spirit to some manga reader apps that scrape its content from websites. Hammering a site all at once for updates and downloads on-demand would basically be DoS-ing it and risk getting throttled or blocked. Caching and spreading out the queries is much more friendly.

**Comment by geominorai** [comment]
Adding to the above, caching Tweets also allows for estimating the profile's posting frequency. This allows for the global updates to be adaptive, such as checking once every few days for very infrequent (monthly/bi-monthly) tweeters while those more frequently posting could be checked once every few hours.

This leaves room for a possible priority updates category users may be able to configure for very important profiles like storm and earthquake alerts (if they have not already left Twitter).

**Comment by geominorai** [comment]
The bonus for this approach is that it is just piggybacking on the individual profile feeds being able to load. There is no extra endpoint to manage.

In the event there is downtime that requires an app update, the feed can try to slowly repopulate missed Tweets from the profile's last checked timestamp once the user installs the updated apk. If there are too many posts to load at a time, it may take a few global updates to finish fetching everything, but users should be understanding due to the rate limits.

**Comment by summersab** [comment]
@jonjomckay - since the feed doesn't (and likely won't) work, would it be possible to trigger a search for the trending topic/hashtag on a search engine when an item is clicked? That way, users can at least figure out why something is trending - that's good enough for me.

It would be great if you could provide the following:
- Option to change search engine (include the most common ones, maybe allow the user to provide their own search URL)
- Option to open the search in the default browser or in a WebView tab within Fritter

The second functionality isn't as important, but it would be nice to have.

Thank you for your hard work!

**Comment by eye-writ** [comment]
Don't know why, but the feed function is working for me this morning.

Version v3.0.0-beta7+300000957

**Comment by TheHCJ** [comment]
My feed is also working now on beta7

**Comment by AndiDink00x** [comment]
@TheHCJ @eye-writ  yeah until <null type string> error shows up , thats the case on my part

**Comment by eye-writ** [comment]
> @TheHCJ @eye-writ yeah until error shows up , thats the case on my part

After a while I get the error too, but the old workaround of clearing the app's storage data still works.

**Comment by TheHCJ** [comment]
I am working on getting the feed and tweet search working again, feel free to check out the repo below
[Repo](https://github.com/TheHCJ/fritter)

**Comment by AndiDink00x** [comment]
So far so good in terms of functionality but I'm experiencing some really annoying freezes and lag while scrolling the feed

**Comment by TheHCJ** [comment]
Going to fix that later today 😃

**Comment by eye-writ** [comment]
> I am working on getting the feed and tweet search working again, feel free to check out the repo below [Repo](https://github.com/TheHCJ/fritter)

Sorry to be a divvy, but how do I install your fix? 😕

**Comment by ha-gorem** [comment]
> > I am working on getting the feed and tweet search working again, feel free to check out the repo below [Repo](https://github.com/TheHCJ/fritter)
> 
> Sorry to be a divvy, but how do I install your fix? confused

from here
https://github.com/TheHCJ/fritter/releases

**Comment by TheHCJ** [comment]
https://github.com/TheHCJ/fritter/releases/download/apkv1/app-debug.apk

**Comment by AndiDink00x** [comment]
@TheHCJ Found a bug , if default tab is set from blank to Feed , tweets get duplicated

**Comment by moocow1452** [comment]
No releases on your repo and the APK link you gave 404d on me. Thanks for your contributions.

**Comment by eye-writ** [comment]
This link work for you? - https://github.com/TheHCJ/fritter/releases

**Comment by TheHCJ** [comment]
Here is the latest file
[app-debug.apk](https://github.com/jonjomckay/fritter/files/11409216/app-debug.apk.zip)

**Comment by AndiDink00x** [comment]
@TheHCJ thats a new pipe apk

**Comment by TheHCJ** [comment]
> @TheHCJ thats a new pipe apk

Shoot, wrong apk :0

**Comment by AndiDink00x** [comment]
@TheHCJ Damn dude just installed and the app is flawless. Saved the whole app and probably made thousands of its users happy and able to use Fritter again.

**Comment by eye-writ** [comment]
It's a lifesaver of an update from HCJ, but continued love for Jonjo too. Not easy when the Tesla nonce throws a spanner in the works! 🔧

**Comment by TheHCJ** [comment]
I will try to get a pull request in tomorrow to fix this in the main repo

**Comment by Igaru22** [comment]
> @TheHCJ Damn dude just installed and the app is flawless. Saved the whole app and probably made thousands of its users happy and able to use Fritter again.

Which apk are you referring to? I got 404 on all the ones in this thread

**Comment by Igaru22** [comment]
Crashes on open

**Comment by TheHCJ** [comment]
> Crashes on open

Seems the api has broke again, I feel like feeds cannot be fixed until Elon opens up the api or Fritter starts using it's own way of getting data off twitter.

**Comment by eye-writ** [comment]
I'm probably speaking too soon, but it's still working fine for me.

**Comment by AndiDink00x** [comment]
@TheHCJ still works pretty good

**Comment by Igaru22** [comment]
Yesterday it worked on my Android 11 device but crashed on open on my Android 5 tablet. Today neither works

**Comment by AndiDink00x** [comment]
@Igaru22 did you download the correct apk from @TheHCJ  ?

**Comment by moocow1452** [comment]
Beta 7 has been working fine for me with regular storage clears. I haven't
been able to access HCJ's build yet.

**Comment by TheHCJ** [comment]
Now bug testing my version before making a pull request

**Comment by TheHCJ** [comment]
Just made a pull request to fix this, I will attach an apk soon!
https://github.com/jonjomckay/fritter/pull/698

**Comment by Igaru22** [comment]
> @Igaru22 did you download the correct apk from @TheHCJ ?

Yes I tried them all. The Android 5 does not seem to work on any 3.x version. The Android 11 worked with Beta 7 for a while. I'm about to check out the new fix

**Comment by TheHCJ** [comment]
Just added an updated apk in the description of this issue! 😀

**Comment by AndiDink00x** [comment]
@TheHCJ Your new updated has a bug on some retweets 
https://imgur.com/a/WmlDdKh

Reverted to your previous apk and those retweets are shown

**Comment by TheHCJ** [comment]
That's a bug with beta8. Looking into a fix!

**Comment by AndiDink00x** [comment]
@TheHCJ ok your fix needs to be pulled to master branch as fast as possible so this issue can be closed and be able reporting bugs I've found, which are a bunch

**Comment by TheHCJ** [comment]
I would like that to happen too, once this is pulled I will probably start
work on a UI refresh.

**Comment by jonjomckay** [comment]
**TLDR; I'm not going to be reinstating the feed, as it's too hard to maintain and bad for people's mental health.**

Hi all. I won't be reinstating the feed to Fritter. I've been thinking about it a lot this past week, and there are multiple reasons why I don't think it's a good idea. 

As I mentioned above, maintenance of the feed is by far the largest time and effort drain when developing Fritter. It's the least fun and the least interesting part of the app for me to work on. It's fragile (out of necessity because there's no direct "give me the feed" API on Twitter), and prone to a lot of breakage, due to Twitter constantly changing their APIs. The endpoint being completely removed recently, then being added back in was the final straw here - I simply don't want to, and can't, maintain something that relies on something so unreliable. I don't like it, and I know you all don't like it when you get cryptic error messages for days on end - it's almost never because of me, it's Twitter breaking things!

Alongside that, removing the feed was a bit of a cathartic moment for me. Since I took it out, I'm able to think about Fritter without feeling frustrated or overwhelmed. I don't mean this in a negative way, but some of the comments on recent issues and Play Store reviews are starting to bring those feelings back, and I'd rather not have that looming over my head.

In my head, Fritter was never intended to be another one of those endless scroll social media apps. I've never been a fan of doomscrolling, and the mental health issues it can lead to or exacerbate are not something that I would like to encourage. GitHub issue peer pressure is the true killer 🙂

Fritter started off as a fun project, to create a privacy friendly viewer for Twitter (well, Nitter, technically), and I'd like to take it back to that - something fun and useful for me to create, and for the user to enjoy.

For those who disagree with my decision, or don't respect it, I would really encourage you to research the mental dangers of doomscrolling and endless social media feeds. I know a lot of you have it under control, and others don't want to be told what to do, but I would really encourage and appreciate it. There's a very good accessible article [from the BBC on it here](https://www.bbc.com/worklife/article/20210226-the-darkly-soothing-compulsion-of-doomscrolling).

@TheHCJ if you'd like to continue a fork that has the feed, I'm happy for you to do so (MIT license, bro), but we should chat about your branding, changing the use of the Fritter name, choosing a new package name, etc. It would be nice to share a base, as it doesn't pay to write the same things twice 😅

Thank you all for understanding, and if you'd like to continue this conversation, I'm going to open a discussion for it, rather than keep this endless thread (heh).
